### PR TITLE
#18309 Owner user should have permission to edit contentlet

### PR DESCRIPTION
**Pull request by freddyucv**
The lock permission error was show because the user not have EDIT_PERMISSIOn and after save a contentlet a unlock action is called, The fix was just add a if if the content is not lock it no need unlock it

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 1]
```diff
@@ -330,6 +333,94 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
+        final ContentType contentType = new ContentTypeDataGen().nextPersisted();
+        final Contentlet contentlet = new ContentletDataGen(contentType.id()).next();
+
+        final ESContentletAPIImpl esContentletAPI = new ESContentletAPIImpl();
+        final Contentlet contentletSaved = esContentletAPI.checkin(contentlet, APILocator.systemUser(), false);
+
+        addPermission(role, contentType, PermissionLevel.WRITE);
+        addPermission(role, contentletSaved, PermissionLevel.WRITE);
+
+        esContentletAPI.lock(contentletSaved, user, false);
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:377]: We can add an assert checking that the content is actually locked

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 2]
```diff
@@ -330,6 +333,94 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
+        final User user = new UserDataGen().roles(role).nextPersisted();
+        final ContentType contentType = new ContentTypeDataGen().nextPersisted();
+        final Contentlet contentlet = new ContentletDataGen(contentType.id())
+                 .next();
+
+        addPermission(role, contentType, PermissionLevel.WRITE);
+
+        final ESContentletAPIImpl esContentletAPI = new ESContentletAPIImpl();
+        final Contentlet contentletSaved = esContentletAPI.checkin(contentlet, user, false);
+        esContentletAPI.lock(contentletSaved, user, false);
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:400]: same as previous comment

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 3]
```diff
@@ -330,6 +333,94 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
+
+            throws DotDataException, DotSecurityException {
+
+        APILocator.getPermissionAPI().save(
+                getPermission(role, contentType, permissionLevel.getType()),
+                contentType, APILocator.systemUser(), false);
+    }
+
+    @NotNull
+    private Permission getPermission(Role role, Permissionable permissionable, int permissionPublish) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:416]: Codacy Issue found: [Parameter 'role' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 4]
```diff
@@ -330,6 +333,94 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
+    /**
+     * Method to test: {@link ESContentletAPIImpl#lock(Contentlet, User, boolean)}
+     * Given Scenario: The contentlet's owner without permission to EDIT try to lock the contebtlet
+     * ExpectedResult: Should work
+     *
+     * @throws DotSecurityException
+     * @throws DotDataException
+     */
+    @Test()
+    public void whenOwnerTryToLockShouldWork() throws DotSecurityException, DotDataException {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:389]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 5]
```diff
@@ -45,8 +44,12 @@
 import com.dotmarketing.util.WebKeys.Relationship.RELATIONSHIP_CARDINALITY;
 import com.liferay.portal.model.User;
 import com.liferay.util.StringPool;
+
+import java.util.ArrayList;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:48]: Codacy Issue found: [Avoid unused imports such as 'java.util.ArrayList'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 6]
```diff
@@ -20,14 +20,12 @@
...
-import com.dotcms.datagen.TestDataUtils;
+import com.dotcms.datagen.*;
 import com.dotcms.util.CollectionsUtils;
 import com.dotcms.util.IntegrationTestInitService;
 import com.dotmarketing.beans.Host;
-import com.dotmarketing.business.APILocator;
-import com.dotmarketing.business.CacheLocator;
-import com.dotmarketing.business.RelationshipAPI;
+import com.dotmarketing.beans.Permission;
+import com.dotmarketing.business.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:28]: Codacy Issue found: [Avoid unused imports such as 'com.dotmarketing.business'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 7]
```diff
@@ -37,6 +35,7 @@
 import com.dotmarketing.portlets.folders.business.FolderAPI;
 import com.dotmarketing.portlets.languagesmanager.business.LanguageAPI;
 import com.dotmarketing.portlets.languagesmanager.model.Language;
+import com.dotmarketing.portlets.rules.model.Rule;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:38]: Codacy Issue found: [Avoid unused imports such as 'com.dotmarketing.portlets.rules.model.Rule'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 8]
```diff
@@ -20,14 +20,12 @@
 import com.dotcms.contenttype.model.type.ContentTypeBuilder;
 import com.dotcms.contenttype.model.type.SimpleContentType;
 import com.dotcms.contenttype.transform.contenttype.StructureTransformer;
-import com.dotcms.datagen.ContentletDataGen;
-import com.dotcms.datagen.TestDataUtils;
+import com.dotcms.datagen.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:23]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 9]
```diff
@@ -330,6 +333,94 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
+    /**
+     * Method to test: {@link ESContentletAPIImpl#lock(Contentlet, User, boolean)}
+     * Given Scenario: A user with {@link PermissionLevel#EDIT} permission try to lock a contentlet
+     * ExpectedResult: Should work
+     *
+     * @throws DotSecurityException
+     * @throws DotDataException
+     */
+    @Test
+    public void whenTryToLockShouldWork() throws DotSecurityException, DotDataException {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:365]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java [hunk 10]
```diff
@@ -310,6 +310,13 @@ public boolean doesUserHavePermission(final Permissionable permissionable, int p
 
 		final List<Permission> perms =  getPermissions(permissionable, true);
 		final boolean isContentlet = permissionable instanceof Contentlet;
+
+		if (isContentlet &&
+				PermissionLevel.WRITE.getType() == permissionType &&
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java:315]: does this apply only for `WRITE` level or any other also?

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+import com.tngtech.java.junit.dataprovider.DataProvider;
+import com.tngtech.java.junit.dataprovider.DataProviderRunner;
+import com.tngtech.java.junit.dataprovider.UseDataProvider;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Map;
+
+import static com.dotcms.util.CollectionsUtils.map;
+import static org.junit.Assert.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:32]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+@RunWith(DataProviderRunner.class)
+public class CheckinContentActionletTest {
+
+    private static class TestCase {
+        Contentlet contentlet;
+        User user;
+
+        boolean hasWritePermission;
+
+        public TestCase(Contentlet contentlet, User user, boolean hasWritePermission) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:45]: Codacy Issue found: [Parameter 'user' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 13]
```diff
@@ -330,6 +332,115 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:364]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
+package com.dotmarketing.portlets.workflows.actionlet;
+
+import com.dotcms.contenttype.model.field.*;
+import com.dotcms.contenttype.model.type.ContentType;
+import com.dotcms.datagen.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:5]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java [hunk 15]
```diff
@@ -0,0 +1,150 @@
+package com.dotmarketing.portlets.workflows.actionlet;
+
+import com.dotcms.contenttype.model.field.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:3]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.contenttype.model.field'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 16]
```diff
@@ -330,6 +332,115 @@ public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
...
+    /**
+     * Method to test: {@link ESContentletAPIImpl#lock(Contentlet, User, boolean)}
+     * Given Scenario: The contentlet's owner without permission to EDIT try to lock the contebtlet
+     * ExpectedResult: Should work
+     *
+     * @throws DotSecurityException
+     * @throws DotDataException
+     */
+    @Test()
+    public void whenOwnerTryToLockShouldWork() throws DotSecurityException, DotDataException {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:399]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    private static void addWritePermissionToContentlet(final Role role, final Contentlet contentlet) throws DotDataException {
+        final Permission permission = getPermission(role, contentlet, PermissionLevel.EDIT.getType());
+
+        try {
+            APILocator.getPermissionAPI().save(permission, contentlet, APILocator.systemUser(), false);
+
+        } catch (DotDataException | DotSecurityException e){
+            throw new RuntimeException(e);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:92]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 18]
```diff
@@ -20,19 +20,18 @@
 import com.dotcms.contenttype.model.type.ContentTypeBuilder;
 import com.dotcms.contenttype.model.type.SimpleContentType;
 import com.dotcms.contenttype.transform.contenttype.StructureTransformer;
-import com.dotcms.datagen.ContentletDataGen;
-import com.dotcms.datagen.TestDataUtils;
+import com.dotcms.datagen.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:23]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 19]
```diff
@@ -20,19 +20,18 @@
...
 import com.dotcms.util.CollectionsUtils;
 import com.dotcms.util.IntegrationTestInitService;
 import com.dotmarketing.beans.Host;
-import com.dotmarketing.business.APILocator;
-import com.dotmarketing.business.CacheLocator;
-import com.dotmarketing.business.RelationshipAPI;
+import com.dotmarketing.beans.Permission;
+import com.dotmarketing.business.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:28]: Codacy Issue found: [Avoid unused imports such as 'com.dotmarketing.business'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

**Review by dsilvam — APPROVED** [review]

**Review by jdotcms — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 18 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:23]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java (hunk 17 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:92]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java (hunk 12 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:45]: Codacy Issue found: [Parameter 'user' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 13 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:364]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 19 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:28]: Codacy Issue found: [Avoid unused imports such as 'com.dotmarketing.business'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java (hunk 11 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:32]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 16 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java:399]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java (hunk 15 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:3]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.contenttype.model.field'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java (hunk 14 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/CheckinContentActionletTest.java:5]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5365897)

**Pull request merged**